### PR TITLE
disallow setting parameters not registered in validParams

### LINKED ISSUE #10357: Don't allow developers to set non-existent parameters
## Rationale

Retrieving un advertised parameter is very bad practice yet it's occurring in some of the modules regularly. This makes these objects brittle. This should be prevented. This can be done by putting additional logic in `Parameters::get()` and/or `MooseObject::getParam` (probably the former).

## Description

We are seeing objects in SM and TM where users are retrieving parameters with the normal `getParam` interface, but those parameters are not added in the `validParams` method (or "inherited"). This creates brittle objects and tie them to specific blocks of code that must know which parameters to set before constructing those objects. For an example see `modules/solid_mechanics/src/kernels/StressDivergenceRZ.C:_volumetric_locking_correction`

## Impact

Unknown: This change will definitely break existing code (see example above) but we can use the results of proposing this change to fix up issues and then prevent them from occurring in the future.

### PR DESCRIPTION
**Pull request by rwcarlsen**
There were several places in moose and the modules where we were inserting parameters that weren't logged in validParams functions that had to be fixed.  All setting of the ``_fe_problem`` special parameter was also consolidated/refactored to reside inside the factory.

fixes #10357

### CODE REVIEW

**Comment by permcody** [comment]
Closing due to inactivity - Note: I definitely want this in the framework when you have time. Looks like you'll need to rebase and squash anyway.

**Comment by rwcarlsen** [comment]
Picking this back up.

**Comment by rwcarlsen** [comment]
ping - this should be ready. @permcody

File: framework/src/actions/CreateProblemAction.C [hunk 1]
```diff
@@ -47,6 +47,15 @@ validParams<CreateProblemAction>()
...
                                        "grab the latest file available)");
 
+#ifdef LIBMESH_HAVE_PETSC
+  params.addParam<MultiMooseEnum>(
+      "petsc_options", MultiMooseEnum("", "", true), "TODO: add doc message");
+  params.addParam<std::vector<std::string>>(
+      "petsc_inames", std::vector<std::string>(), "TODO: add doc message");
+  params.addParam<std::vector<std::string>>(
+      "petsc_values", std::vector<std::string>(), "TODO: add doc message");
+#endif
```
Comment (permcody) [inline framework/src/actions/CreateProblemAction.C]: Hmm, this isn't quite right. These parameters should be added through this method:
https://github.com/idaholab/moose/blob/devel/framework/src/utils/PetscSupport.C#L758-L803

**Review by permcody — CHANGES_REQUESTED** [review]
This looks good! Let's figure out why the internal app is failing, and fixup the PETSc options.

**Review by permcody — APPROVED** [review]

**Pull request merged**